integration-tests/deployment/multiclient.go

```diff
@@ -95,30 +99,72 @@ func (mc *MultiClient) CodeAt(ctx context.Context, account common.Address, block
 
 func (mc *MultiClient) NonceAt(ctx context.Context, account common.Address) (uint64, error) {
 	var count uint64
-	err := mc.retryWithBackups(func(client *ethclient.Client) error {
+	err := mc.retryWithBackups("NonceAt", func(client *ethclient.Client) error {
 		var err error
 		count, err = client.NonceAt(ctx, account, nil)
 		return err
 	})
 	return count, err
 }
 
-func (mc *MultiClient) retryWithBackups(op func(*ethclient.Client) error) error {
+func (mc *MultiClient) WaitMined(ctx context.Context, tx *types.Transaction) (*types.Receipt, error) {
+	mc.lggr.Debugf("Waiting for tx %s to be mined", tx.Hash().Hex())
+	// no retries here because we want to wait for the tx to be mined
+	resultCh := make(chan *types.Receipt)
+	doneCh := make(chan struct{})
+
+	waitMined := func(client *ethclient.Client, tx types.Transaction) {
+		mc.lggr.Debugf("Waiting for tx %s to be mined with client %v", tx.Hash().Hex(), client)
+		receipt, err := bind.WaitMined(ctx, client, &tx)
+		if err != nil {
+			mc.lggr.Warnf("WaitMined error %v with client %v", err, client)
+			return
+		}
+		select {
+		case resultCh <- receipt:
+		case <-doneCh:
+			return
+		}
+	}
+
+	for _, client := range append([]*ethclient.Client{mc.Client}, mc.Backups...) {
+		txn := tx
+		c := client
+		go waitMined(c, *txn)
+	}
+	var receipt *types.Receipt
+	select {
+	case receipt = <-resultCh:
+		close(doneCh)
+		return receipt, nil
+	case <-ctx.Done():
+		mc.lggr.Warnf("WaitMined context done %v", ctx.Err())
+		time.Sleep(20 * time.Second)
```

Comment:
probably a wait group would be cleaner here? then you can just wg.Wait() until they all return

Reply:
sleep was leftover hacking in the early stage. no need to wait; removed.